### DISCUSSION TITLE
Documents with distinct DOIs marked as duplicate

### DISCUSSION THREAD

**Discussion by awa5114** [papis/papis#935] [Q&A]
I tried running the following commands in succession:

```
papis add https://doi.org/10.1029/WR022i010p01397
papis add  https://doi.org/10.1029/WR022i010p01404
```

I got a duplicate warning. While the two documents are indeed very similar, they are definitely not duplicate and have distinct DOIs. Furthermore I tried manually resetting the `ref:` key for the first document and then clearing the cache to see if that was causing the issue but I still get the error.

Why are the above documents being marked as duplicate when I try to add the second one? Please help.

**Comment by alexfikl** [comment]
Hm, to check for duplicates, we use the `unique-document-keys` setting
https://papis.readthedocs.io/en/latest/configuration.html#confval-unique-document-keys
which does contain the DOI and the ref and things, so it should definitely not mark those two documents as duplicates.

Have you checked the `info.yaml` files to see if, for some reason, it downloaded the same metadata for those keys? 

Also, if you just give it the URL, I'm not sure it sets the DOI. Does it? I would recommend trying
```
papis add --from doi '10.1029/WR022i010p01397'
papis add --from doi '10.1029/WR022i010p01404'
```
to see if it makes any difference.

**Reply by alexfikl** [reply]
Hm, this seems to be some sort of bug. I can't get the direct call to work locally. Will have to investigate!

EDIT: At least one bug #936..

**Comment by awa5114** [comment]
@alexfikl I did a bit of my own debugging, and I think I've found the issue in `utils.locate_document_in_lib()`. That function loops through the `unique-document-keys` in the library configuration and queries the database for each key. If a key is found, it returns a document, which means we end up in the `else` block one level up in the stack (`add.run()`) and get the duplicate document warning.  [By default](https://papis.readthedocs.io/en/latest/configuration.html#confval-unique-document-keys), the `ref` key is contained within the `unique-document-keys` configuration setting.

The same reference is generated for the document because the metadata for the actual publication is similar enough that the [`ref-format`](https://papis.readthedocs.io/en/latest/configuration.html#confval-ref-format) results in the same reference key. What makes things worse is that the database query mechanism also matches partial strings, so even if I manually change the reference of the first document to something else using for example

`papis update --set ref Urban_Water_Bal_Grimmo_1986_1 "Urban Water Balance"`

`docs = db.query_dict({"ref": "Urban_Water_Bal_Grimmo_1986"})` will still match a document with a `ref` of `Urban_Water_Bal_Grimmo_1986_1`, so that won't work.

The only viable workaround I actually found was to manually override the ref when importing from the DOI as follows, accepting the merge changes and adding the document:

`papis add --from doi '10.1029/WR022i010p01404' --set ref Urban_Water_Bal_Grimmo_2`

But it still feels like the matching algorithm could be improved. Shouldn't *all* the `unique-document-keys` match for a document to be marked as duplicate? And then if it is not a duplicate, wouldn't it make sense to then programmatically modify the reference such that the two documents do not clash?

**Reply by alexfikl** [reply]
Hm, I think the main issue here is that the `ref` is not as unique as it pretends to be. We make no effort to ensure that it is when adding new documents (like you're seeing). There's a `papis doctor` check that can be used for that (`duplicated-keys`), but it's not enabled by default either.

> Shouldn't all the unique-document-keys match for a document to be marked as duplicate?

Maybe as a start we can just remove the `ref` from the default keys, since it doesn't really work for that? The remaining items on that lists should be pretty good as unique identifiers, e.g. if the DOI matches, there really isn't any point in looking at anything else.

> wouldn't it make sense to then programmatically modify the reference such that the two documents do not clash?

I'm not sure about this part. The `ref` is not really meant to be a unique identifier. It's mostly used when exporting to BibTeX and that part of the code should probably make sure that the refs are unique for that file.

**Reply by alexfikl** [reply]
Also, the duplicate warning isn't supposed to be that scary :grin: If you know that your documents are different, just breeze through it!

**Reply by alexfikl** [reply]
@amine-aboufirass Can you check if #939 helps with this? It's merged in `main` now.
